### ISSUE TITLE
Client only properties being wiped

### ISSUE THREAD

**Issue by STAH** [OpenRIAServices/TestIssues#108]
<sub>This issue was imported from [CodePlex](http://openriaservices.codeplex.com/workitem/9)</sub>

**[ColinBlair](https://github.com/ColinBlair)** <sup>wrote 2013-10-03 at 20:12</sup>
During a load or submit, an entity that has client side only properties attached to it through a partial class will have those properties wiped by the load. This is caused because the load from the server is first placed into a new entity object which is then cloned into the existing entity.

**Comment by STAH** [comment]
**[koimad](https://github.com/koimad)** <sup>wrote 2013-10-24 at 02:23</sup>
This could possibly be tip of the issue there is many facets to this issue with Entity Tracking do we Track client side property changes ? do they make the entity Dirty ? and so forth. The Extraction code clearly looks at all properties to see if they are required.  
  
Code could be altered to only set the MetaMember.IsDataMember if the data member attribute is applied to the property, but this would restrict monitoring client side only properties all together making the code very rigid.  
  
From a Developer that has come across this problem in the past and worked around it using a property with a getter and a method for setting it, I would prefer a simpler solution. Make the IsMergingState property on the Entity Class protected internal, this then can be used within the setter of the property to check to see if a merge is occurring or not before setting the value and raising and events required to be raised. This solution would clearly be a very quick solution, other solutions could be explored as more issues are raised about the behaviour required of client side properties in the future.

**Comment by STAH** [comment]
**[ColinBlair](https://github.com/ColinBlair)** <sup>wrote 2013-10-24 at 03:32</sup>
The path I was looking at was tracking what properties came from the server and merging only those properties. I think that solution will work better for the Web API version.

**Comment by STAH** [comment]
**[koimad](https://github.com/koimad)** <sup>wrote 2013-10-31 at 02:02</sup>
Colin I've made a start in my fork, ([https://git01.codeplex.com/forks/koimad/riaservicesclientproperties](https://git01.codeplex.com/forks/koimad/riaservicesclientproperties))  
I think I've covered most things and in my own little test project client side properties are now preserved for properties that are not streamed from the server.  
  
I'm about to start updating the tests to support my changes but I have a problem with the Desktop tests, in that when I run them they don't run, just go to pending. Is this where you have got to with the Desktop tests, or have I got a problem with my laptop?  
  
Brian.

**Comment by STAH** [comment]
**[ColinBlair](https://github.com/ColinBlair)** <sup>wrote 2013-10-31 at 02:17</sup>
Do you have the Northwind and Adventureworks databases setup? Do what you can for the unit tests, I have had offers from others to work on the tests (possibly including a move to XUnit) but they are not 100% at the moment.

**Comment by STAH** [comment]
**[koimad](https://github.com/koimad)** <sup>wrote 2013-10-31 at 02:37</sup>
I haven't got any databases setup at the moment, I'm looking at giving myself a fresh windows 8 install and VS2013 and sqlexpress to get the database up and running, at the moment I've just got the sql developer edition on my lap top.

**Comment by STAH** [comment]
**[ColinBlair](https://github.com/ColinBlair)** <sup>wrote 2013-10-31 at 06:55</sup>
I was taking a look at what is checked in. I don't think Deserializing is the correct term. The issue we are looking at is actually merging, not deserializing. Which brings up the point that IsMergingState is already part of the entity, it is just marked internal. If IsMergingState is made protected instead of internal then see if that gets us where we need to be without adding something new.  
  
In MetaMember I am thinking IsMergable instead of IsSerialized. I don't see yet how you are planning to populate IsSerialized/IsMergable. Perhaps a NotMergableAttribute?  
  
In the ObjectStateUtility, your LINQ query would be a performance problem. Remove the LINQ namespace from ObjectStateUtility and instead use a separate cached list of MergableProperties in the MetaType.

**Comment by STAH** [comment]
**[koimad](https://github.com/koimad)** <sup>wrote 2013-10-31 at 10:17</sup>
Colin, I'd did comment above on using the IsMergingState to control whether a property is reset or not, but this would be up to the developer of the client side property to implement the behaviour in the setter to check IsMerging before setting the member variable or not.  
You said you was thinking of using whether the property came from the server, which is what I implemented, In the OnDeserialising method I acciedently added a blank line. My implementation uses the fact that this method which is decorated with the attribute so the system code invokes this code when the entity is being streamed from the server an sets the member isdeserialising. In the property changing method I simply look at isdeseialising to see if the. Property is being set from the server. If it is I mark the property IsSerialised, I added on the MetaMember class to say that it came from the server. When state is extracted I use this property to see if a property should be extracted or not and hence merged.  
  
My implementation has some draw backs in that what happens if you wanted to have a property on both client server that was only set in the client and passed back to the server, never streamed from the server. It also effects what properties that are tracked for entity state behavior. Form my comment earlier it may be better for now to go for the simple implementation I suggested for now, to provide a short term solution to the issue and wait until a clearer picture of exactly what behaviour is required from client side only properties that also include modified state behaviour of the entity.  
  
I was intending to make the IsMergingState property internal protected so that it could also be accessed in other client side class methods to control behavior if required.  
  
Brian
